crates/conjure_core/src/rules/tuple.rs

```diff
@@ -0,0 +1,303 @@
+use conjure_core::ast::Expression as Expr;
+use conjure_core::ast::SymbolTable;
+use conjure_core::rule_engine::{
+    register_rule, ApplicationError::RuleNotApplicable, ApplicationResult, Reduction,
+};
+use itertools::Itertools;
+
+use crate::ast::Atom;
+use crate::ast::Domain;
+use crate::ast::Expression;
+use crate::ast::Literal;
+use crate::ast::Name;
+use crate::into_matrix_expr;
+use crate::matrix_expr;
+use crate::metadata::Metadata;
+use crate::rule_engine::ApplicationError;
+
+//TODO: largely copied from the matrix rules, This should be possible to simplify
+#[register_rule(("Base", 2000))]
+fn index_tuple_to_atom(expr: &Expr, symbols: &SymbolTable) -> ApplicationResult {
+    // i assume the MkOpIndexing is the same as matrix indexing
+    let Expr::SafeIndex(_, subject, indices) = expr else {
+        return Err(RuleNotApplicable);
+    };
+
+    let Expr::Atomic(_, Atom::Reference(Name::WithRepresentation(name, reprs))) = &**subject else {
+        return Err(RuleNotApplicable);
+    };
+
+    if reprs.first().is_none_or(|x| x.as_str() != "tuple_to_atom") {
+        return Err(RuleNotApplicable);
+    }
+
+    let repr = symbols
+        .get_representation(name, &["tuple_to_atom"])
+        .unwrap()[0]
+        .clone();
+
+    let decl = symbols.lookup(name).unwrap();
+
+    let Some(Domain::DomainTuple(_)) = decl.domain().cloned().map(|x| x.resolve(symbols)) else {
+        return Err(RuleNotApplicable);
+    };
+
+    let mut indices_are_const = true;
+    let mut indices_as_lits: Vec<Literal> = vec![];
+
+    for index in indices {
```

Comment:
Can tuples be multi-dimensional (which would mean they could be indexed by x[1,2]), or are they always one dimensional (x[1])?

If they are always one dimensional, this could be simplified further.